Support for save/load on hash internal state to allow resumable hashing

thenickdude opened the pull request:
In my project I'm hashing some very large files using AWS Lambda, which has a fixed maximum execution time. So partway through the file when I run out of execution time, I need to be able to save the state of the hash to disk, start a new Lambda, and resume that state in the new Lambda to continue hashing the file.

So far I have implemented this just for SHA1/CRC32, which are the hashes I'll be using.

Is this a concept that you're interested in merging? If so I could add support for more/all hashes.

Daninet commented on src/sha1.c:
```diff
@@ -176,6 +176,32 @@ void Hash_Final() {
 }
 
+WASM_EXPORT
+void Hash_Save() {
```
With a `WASM_EXPORT uint8_t* Hash_GetState() { return context; }` function it would be more simple.

Daninet commented on lib/blake2s.ts:
```diff
@@ -112,8 +112,11 @@ export function createBLAKE2s(
         },
       update: (data) => { wasm.update(data); return obj; },
       digest: (outputType) => wasm.digest(outputType) as any,
+      save: () => wasm.save() as any,
```
```suggestion
      save: () => wasm.save(),
```

Daninet commented on lib/WASMInterface.ts:
```diff
@@ -142,6 +154,30 @@ export async function WASMInterface(binary: any, hashLength: number) {
+      throw new Error('save() called before init()');
+    }
+    if (stateLength === 0) {
+      throw new Error('save() not supported for this hash');
```
It would be good to implement it in all algorithms before release.

Daninet commented on lib/WASMInterface.ts:
```diff
@@ -29,6 +29,14 @@ export type IHasher = {
+   */
+  save: () => Uint8Array;
+  /**
+   * Resume a state that was created by save(). init() must have been called first
```
```suggestion
   * Resume a state that was created by save().
```

Daninet commented on lib/WASMInterface.ts:
```diff
@@ -142,6 +154,30 @@ export async function WASMInterface(binary: any, hashLength: number) {
+    return memoryView.slice(0, stateLength);
+  };
+
+  const load = (data: IDataType) => {
```
```suggestion
  const load = (data: Uint8Array) => {
```
It would be also nice to check if the `data.length === stateLength`.

Daninet reviewed (commented):
I like the idea of having this feature, although I think it would be much easier to implement it by returning a pointer to the state object from WASM to JS. Something similar to `Hash_GetBuffer()` in hash-wasm.h. And after that, WASMInterface could take over the responsibility of slicing and rewriting the memory region corresponding to the hash state.

thenickdude commented:
> after that, WASMInterface could take over the responsibility of slicing and rewriting the memory region corresponding to the hash state.

Is that possible? I think the state object will have its own internal padding and alignment which would be hard for the JavaScript code to predict? (This is why I manually serialised each field instead of writing the entire struct in one go)

thenickdude commented:
Could the padding bytes be uninitialised memory which leaks state from some previous execution though? It seems safer not to include any padding bytes in the serialised form.

Daninet commented:
WASM memory is always initialized to zero before the instance is created. Padding remains padding even during execution so it cannot happen that a previous hash uses a different padding for the same struct. If somebody is concerned about leaking out state then he destroy all references to the library instance in JS. That will destroy the whole WASM instance including with the associated memory area.

thenickdude commented:
Makes sense. I'll update the PR with these suggestions and start adding more supported hashes

thenickdude commented:
Is this new interface looking good now?

The C code now supplies the size of its state with an exported "STATE_SIZE" constant, and returns a pointer to a buffer of this size from a new `Hash_GetState()` function. WASMInterface takes care of the rest. (Still only implemented for SHA1 and CRC32 so far)

Daninet commented:
Yeah, good job! 👍 
I'm waiting for the rest of the algorithms before merging it.
Maybe the unit tests could be generalized a bit and moved to "api.test.ts". There is an example which cycles over all "createXXX" functions.

thenickdude commented:
Okay, I've implemented the other hashes too now, moved the tests into a pair of generic tests in api.test.ts, and added a section to the Readme.

The C hash functions for Whirlpool, xxhash32 and xxhash64 have had their loose global state variables moved into a struct. For xxhash64 I double checked using the benchmarking tool that this didn't have any negative impact.

`createHMAC()` still throws when .save() or .load() is called on it. We might not want to support these operations on this one because the saved state can leak the signing key, which would be a disaster.

Daninet reviewed (approved):
Awesome, thank you. I will try to release it during the weekend.

thenickdude commented:
We're not supporting compatibility between states saved on different builds of hash-wasm, which makes sense because otherwise you'd have to nail down the internal state of the hash functions, and then the implementations couldn't be improved.

But I just realised that this will prevent me from deploying a new version of my function unless I can be absolutely sure there are no jobs currently in progress, because updated .load()ers could silently give the wrong hashes if they read a state that was still in flight after being written by an old .save() version.

So I'm thinking I should add a hash of the WASM directory to the saved state, which is then verified on state load? This way we can reliably throw an exception when trying to load a state created by a different build. The hash can be calculated during the build step, so it won't really slow this down.

thenickdude commented:
I've added a commit which implements this compatible-state-verification feature now.

Daninet commented:
I believe this detection could be improved.
It's very unlikely that the alignment of the fields will change between versions so I think it's too restrictive to bail out the state loading every time when there is a smaller change in one of the binaries. To address the issue, I suggest generating a hash of the state, individually for each algorithm. `make_json.js` should be able to add a new field to output JSON files containing the hash of the saved state after hashing a long predefined constant.
Also, to keep things efficient I think it's more than enough to use the first 4 bytes of the SHA1 hash output instead of the full hash.

thenickdude commented:
It would be relatively easy in that situation to end up with state bytes that never get touched during hash of the predefined constant, for example the high bytes of length fields. These bytes could be filled with undefined padding values in one hash version, then in a later hash version which increases the size of one of those fields (from uint32_t to uint64_t, say) those uninitialised padding values could suddenly affect the behaviour of the hash without being detected by your scheme.

thenickdude commented:
I think a better idea would be to investigate reproducible-build flags for Clang so the WASM binaries don't change if the binary doesn't get edited, and then go for the per-function 4-byte hash idea you suggested.

Daninet commented:
Yeah, it would be hard to come up with a good sequence of predefined constants, especially with more complicated hash functions like BLAKE3.
What about hashing the WASM binaries individually then? Hashing all files together does not make sense to me, because it's more complex but not really different from using the package version from `package.json` as a differentiator.

thenickdude commented:
>Hashing all files together does not make sense to me, because it's more complex but not really different from using the package version from package.json as a differentiator.

Hashing the files accounts for people who build hash-wasm themselves with different Clang versions, generating different output without bumping the package version.

Daninet commented:
> Hashing the files accounts for people who build hash-wasm themselves with different Clang versions, generating different output without bumping the package version.

Good point.
I think we both agree that hashing the binaries individually is the best solution. :)

Daninet commented:
With the current clang flags + docker-based setup the builds should be already reproducible. This was an important feature from the beginning to have a transparent and reproducible build process.

thenickdude commented:
Okay, I'll work on including the per-function code hashes.

thenickdude commented:
I've implemented this now!
